ouroboros-consensus/docs/report/chapters/future/genesis.tex

```diff
@@ -1499,3 +1500,178 @@ \subsection{Possible optimisations}
 to not being up to date would require careful thought.
 
 \end{itemize}
+
+\section{Suboptimal chain selection}
+
+The Ouroboros papers assume, by way of simplification, that nodes broadcast
+entire chains to each other. Every node collects all of these chains, and
+every slot selects the best chain they saw. In addition, the Ouroboros analysis
+assumes a constant $\Delta$, such that chains are submitted to all nodes in
+the network within $\Delta$ slots. A consequence of $\Delta > 1$ is that
+a newly forged block might not make it to the next slot leader in time; when
+this happens, that next slot leader will forge their block on a different tip
+instead, resulting in a short fork. As long as the chain propagation delay
+stays within $\Delta$, however, the Ouroboros analysis guarantees that
+consensus will nonetheless be achieved.
+
+In the implementation, each node follow the chains of its upstream peers as
+chains of \emph{headers}. These headers provide us with sufficient context to do
+chain selection and pick the best chain\footnote{Technically speaking this is a
+preliminary chain selection, because the chain we chose might contain invalid
+blocks. If this is the case, we will notice when we download the blocks and
+attempt to actually adopt the chain in the chain database. We will then
+disconnect from the peer that served us the offending chain, at which point a
+different chain might now be the preferred chain. This introduces a delay in
+selecting the optimal valid chain, but we accept this risk as sufficiently small
+that we do not pre-emptively download blocks from all chains. After all, there
+is little incentive for an attacker with the right to produce blocks to produce
+\emph{invalid} blocks.}. Under normal circumstances this is the \emph{only}
+chain we will download the blocks of and adopt, but occassionally we might want
+to download blocks from chains \emph{other} than the best chain, perhaps because
+downloading that best chain is taking a very long time (it might even be the
+case that the remote node is intentionally trying to stall us, as a form of
```

Comment:
There's something I don't understand in this narrative: it is clear that we might want to chose to download a sub-optimal chain because the best chain is taking a long time to download. But what I don't get is what will we do with this sub-optimal chain. Could it be that the sub-optimal chain ends up being the optimal chain after some time? 

Also I'm assuming that here we already know that a chain is optimal (or sub-optimal) through its headers. It this correct?

Reply:
>  Could it be that the sub-optimal chain ends up being the optimal chain after some time?

After reading the report, it seems the sub-optimal chain could become the optimal chain if we forge a block on top of it. But what about peers that do not forge a node. Do they have to keep on waiting till all the blocks from the optimal chain are provided?

Reply:
The common case is that the suboptimal chain is a short fork, which due to the way the fork is resolved elsewhere (other nodes forging blocks) ultimately becomes the optimal chain.

Reply:
Not necessarily, the example that we've discussed couple of days ago was following:
1. imagine two nodes A & B  and two chains C1 & C2
2. A follows B
3. A is fetching optimal C1 to adopt it, but also (in parallel) fetching suboptimal chain C2
4. time passes and A is going to be selected as a leader for the next block. But before it's his turn he learns that B have build on top of C2 and now C2 is optimal
5. A is starting to switch to C2 (request fetch to block fetcher)
6. because most blocks were pre-fetched already, block fetcher only has to download the recently forged blocks and thus A has higher probability to forge on C2 

If suboptimal chain was not happening then A would learn that C2 is optimal, instruct block fetcher to get its blocks, but because that operation is time consuming, there is a higher probability that A would forge on C1 (not on C2) and essentially wasting resources, forging a block that will be wasted.

Now disclaimers: 
1. this is how I remembered the example that said to be a motivation behind the suboptimal chain selection. This is for optimalization purposes only. It would be cool if @nfrisby or @edsko  could confirm it
2. What I don't seem to understand just yet was why this example was not part of the chapter in the report and instead the example there (motivation) was about optimal chain being not delivered on time (slowed done, presumably as adversarial behaviour). I would be more then happy to revist this during our sync.

Reply:
This example seems like a nice justification for a specific situation in which suboptimal chain selection is beneficial, but ultimately it boils down to the same problem:
> downloading that best chain is taking a very long time
In this example, either B doesn't know at all about C1 or it is also not being able to download it, so we are in the same situation. (Or B itself forged the tip of C2 and therefore adopted it even though for A it is not the optimal one, this is an extreme case)

Reply:
You are correct @Jasagredo Also as pointed out in the meeting yesterday, this example is merely a optimalization. But without suboptimal chain selection we can also get adversarial attacks on the whole chain. I will add those examples to the chapter (learning tex by copy and paste :) )